File: api/routes/simulator.py
from __future__ import annotations

import asyncio
import json
import logging
from typing import Any, Optional

from fastapi import APIRouter, HTTPException, Query, Request, WebSocket, WebSocketDisconnect
from fastapi.responses import JSONResponse, Response

from api.websocket.manager import WebSocketManager
from simulator.models import (
    AIDecisionMode,
    ScenarioType,
    SimSpeed,
    SimStatus,
    SimulatorConfig,
    SimulatorState,
)
from simulator.replay_engine import MarketReplayEngine
from simulator.report_generator import ReportGenerator
from simulator.scenarios import SCENARIOS, generate_scenario_data
from simulator.session_manager import SessionManager
from simulator.simulator_engine import SimulatorEngine

logger = logging.getLogger(__name__)

router = APIRouter()

_engine: SimulatorEngine | None = None
_session_mgr: SessionManager | None = None
_ws_manager: WebSocketManager | None = None
_replay_engine: MarketReplayEngine | None = None


async def _ws_broadcast_state(state: SimulatorState) -> None:
    try:
        await _get_ws().broadcast_to_room("simulator", json.dumps({
            "event": "SIMULATOR_STATE",
            "timestamp": state.current_timestamp or 0,
            "payload": state.to_dict(),
        }))
    except Exception:
        pass


async def _ws_broadcast_trade(trade: Any) -> None:
    try:
        await _get_ws().broadcast_to_room("simulator", json.dumps({
            "event": "SIMULATOR_TRADE",
            "payload": trade.to_dict() if hasattr(trade, "to_dict") else trade,
        }))
    except Exception:
        pass


async def _ws_broadcast_decision(decision: Any) -> None:
    try:
        await _get_ws().broadcast_to_room("simulator", json.dumps({
            "event": "SIMULATOR_DECISION",
            "payload": decision.to_dict() if hasattr(decision, "to_dict") else decision,
        }))
    except Exception:
        pass


async def _ws_broadcast_candle(candle: Any) -> None:
    try:
        await _get_ws().broadcast_to_room("simulator", json.dumps({
            "event": "SIMULATOR_CANDLE",
            "payload": candle.to_dict() if hasattr(candle, "to_dict") else candle,
        }))
    except Exception:
        pass


def _get_engine() -> SimulatorEngine:
    global _engine
    if _engine is None:
        from council.consensus import ConsensusEngine

        council_engine = ConsensusEngine()
        council_engine.register_defaults()
        _engine = SimulatorEngine(
            replay_engine=MarketReplayEngine(),
            report_generator=ReportGenerator(),
            council_engine=council_engine,
        )
        # Registered once against the module-level singleton engine, not per
        # /ws/simulator connection -- broadcast_to_room() already fans out to
        # every socket in the room from a single registration, and these
        # listener lists have no unsubscribe API, so registering per-connect
        # leaked 4 more closures on every reconnect.
        _engine.on_state(lambda s: asyncio.ensure_future(_ws_broadcast_state(s)))
        _engine.on_trade(lambda t: asyncio.ensure_future(_ws_broadcast_trade(t)))
        _engine.on_decision(lambda d: asyncio.ensure_future(_ws_broadcast_decision(d)))
        _engine.on_candle(lambda c: asyncio.ensure_future(_ws_broadcast_candle(c)))
    return _engine


def _get_session_mgr() -> SessionManager:
    global _session_mgr
    if _session_mgr is None:
        _session_mgr = SessionManager()
    return _session_mgr


def _get_ws() -> WebSocketManager:
    global _ws_manager
    if _ws_manager is None:
        _ws_manager = WebSocketManager()
    return _ws_manager


def _init_engine_from_state(state: SimulatorState) -> SimulatorEngine:
    eng = _get_engine()
    eng._replay = MarketReplayEngine()
    eng._replay._candles = list(state.candles)
    eng._replay._index = state.current_candle_index
    eng._replay._symbol = state.config.symbol
    eng._state = state
    return eng


# ── Session CRUD ────────────────────────────────────────────────────────────


@router.get("/simulator/sessions")
def list_sessions():
    mgr = _get_session_mgr()
    sessions = mgr.list_sessions()
    return {"sessions": [s.to_dict() for s in sessions]}


@router.post("/simulator/sessions")
def create_session(config: SimulatorConfig, name: str = ""):
    mgr = _get_session_mgr()
    state = mgr.create(config, name=name)
    return {"session_id": state.session_id, "state": state.to_dict()}


@router.get("/simulator/sessions/{session_id}")
def get_session(session_id: str):
    mgr = _get_session_mgr()
    state = mgr.load(session_id)
    if state is None:
        raise HTTPException(status_code=404, detail="Session not found")
    return state.to_dict()


@router.delete("/simulator/sessions/{session_id}")
def delete_session(session_id: str):
    mgr = _get_session_mgr()
    ok = mgr.delete(session_id)
    if not ok:
        raise HTTPException(status_code=404, detail="Session not found")
    return {"deleted": True}


@router.post("/simulator/sessions/{session_id}/save")
def save_session(session_id: str, name: str = ""):
    eng = _get_engine()
    if eng.state is None or eng.state.session_id != session_id:
        state = _get_session_mgr().load(session_id)
        if state is None:
            raise HTTPException(status_code=404, detail="Session not found")
        _init_engine_from_state(state)
    _get_session_mgr().save(eng.state, name=name)
    return {"saved": True}


@router.get("/simulator/sessions/compare/{id_a}/{id_b}")
def compare_sessions(id_a: str, id_b: str):
    mgr = _get_session_mgr()
    result = mgr.compare_sessions(id_a, id_b)
    if "error" in result:
        raise HTTPException(status_code=404, detail=result["error"])
    return result


# ── Simulation Control ──────────────────────────────────────────────────────


@router.post("/simulator/start")
async def start_simulation(config: SimulatorConfig, name: str = ""):
    eng = _get_engine()
    if eng.running:
        raise HTTPException(status_code=409, detail="Simulation already running")
    sid = await eng.start(config, name=name)
    if eng.state:
        _get_session_mgr().save(eng.state, name=name)
    return {"session_id": sid, "state": eng.state.to_dict() if eng.state else {}}


@router.post("/simulator/pause")
def pause_simulation():
    eng = _get_engine()
    if not eng.state or eng.state.status != SimStatus.RUNNING:
        raise HTTPException(status_code=409, detail="Simulation not running")
    eng.pause()
    return {"status": "PAUSED"}


@router.post("/simulator/resume")
def resume_simulation():
    eng = _get_engine()
    if not eng.state or eng.state.status != SimStatus.PAUSED:
        raise HTTPException(status_code=409, detail="Simulation not paused")
    eng.resume()
    return {"status": "RESUMED"}


@router.post("/simulator/stop")
def stop_simulation():
    eng = _get_engine()
    if eng.state is None:
        raise HTTPException(status_code=409, detail="No active simulation")
    state = eng.stop()
    if state:
        _get_session_mgr().save(state)
    return {"status": "STOPPED", "state": state.to_dict() if state else {}}


@router.post("/simulator/reset")
def reset_simulation():
    eng = _get_engine()
    eng.reset()
    return {"status": "RESET"}


@router.post("/simulator/step")
def step_simulation():
    eng = _get_engine()
    candle = eng.step_candle()
    if candle is None:
        raise HTTPException(status_code=409, detail="Cannot step - simulation not paused or no more data")
    return {"candle": candle.to_dict(), "state": eng.state.to_dict() if eng.state else {}}


@router.get("/simulator/state")
def get_simulator_state():
    eng = _get_engine()
    if eng.state is None:
        return {"status": "IDLE", "state": None}
    return eng.state.to_dict()


@router.post("/simulator/speed")
def set_speed(speed: SimSpeed):
    eng = _get_engine()
    if eng.state is None:
        raise HTTPException(status_code=409, detail="No active simulation")
    eng.state.config.speed = speed
    return {"speed": speed.value}


@router.post("/simulator/seek")
def seek_to_timestamp(timestamp: int):
    eng = _get_engine()
    ok = eng.seek_to(timestamp)
    if not ok:
        raise HTTPException(status_code=400, detail="Invalid timestamp")
    return {"state": eng.state.to_dict() if eng.state else {}}


# ── Trading ──────────────────────────────────────────────────────────────────


@router.post("/simulator/trade")
def manual_trade(
    side: str = Query(pattern="^(BUY|SELL|LONG|SHORT)$"),
    entry_price: float = Query(gt=0),
    stop_loss: float = Query(gt=0),
    take_profit: float = Query(gt=0),
    quantity: float = Query(gt=0),
    leverage: float = Query(1.0, ge=1, le=100),
    trailing_stop: float | None = Query(None, gt=0),
):
    eng = _get_engine()
    if eng.state is None:
        raise HTTPException(status_code=409, detail="No active simulation")
    norm_side = "LONG" if side in ("BUY", "LONG") else "SHORT"
    trade = eng.execute_manual_trade(norm_side, entry_price, stop_loss, take_profit, quantity, leverage, trailing_stop)
    if trade is None:
        raise HTTPException(status_code=400, detail="Failed to create trade")
    return {"trade": trade.to_dict()}


@router.post("/simulator/trade/{trade_id}/close")
def close_trade(trade_id: str, exit_price: float | None = Query(None, gt=0)):
    eng = _get_engine()
    ok = eng.close_trade(trade_id, exit_price)
    if not ok:
        raise HTTPException(status_code=404, detail="Trade not found or already closed")
    return {"closed": True}


@router.post("/simulator/trades/close-all")
def close_all_trades(exit_price: float | None = Query(None, gt=0)):
    eng = _get_engine()
    closed = eng.close_all_trades(exit_price)
    return {"closed_count": closed}


# ── Reports ──────────────────────────────────────────────────────────────────


@router.get("/simulator/report")
def get_report():
    eng = _get_engine()
    report = eng.get_report()
    if report is None:
        raise HTTPException(status_code=409, detail="No active simulation")
    return report.to_dict()


@router.get("/simulator/report/json")
def export_report_json():
    eng = _get_engine()
    data = eng.get_report_json()
    if not data:
        raise HTTPException(status_code=409, detail="No active simulation")
    return Response(content=data, media_type="application/json")


@router.get("/simulator/report/pdf")
def export_report_pdf():
    eng = _get_engine()
    pdf = eng.get_report_pdf()
    if not pdf:
        raise HTTPException(status_code=409, detail="No active simulation")
    return Response(content=pdf, media_type="application/pdf", headers={
        "Content-Disposition": f'attachment; filename="mission_report_{eng.state.session_id}.pdf"',
    })


# ── Scenarios ────────────────────────────────────────────────────────────────


@router.get("/simulator/scenarios")
def list_scenarios():
    return {
        "scenarios": {
            k.value: {
                "name": v["name"],
                "description": v["description"],
            }
            for k, v in SCENARIOS.items()
        }
    }


@router.post("/simulator/scenarios/generate")
def generate_scenario(
    scenario_type: ScenarioType,
    symbol: str = "BTC",
    timeframe: str = "1h",
    num_candles: int = 200,
    start_price: float | None = None,
):
    df = generate_scenario_data(scenario_type, symbol, timeframe, num_candles, start_price)
    if df is None or df.empty:
        raise HTTPException(status_code=500, detail="Failed to generate scenario data")
    candles = []
    for _, row in df.iterrows():
        candles.append({
            "timestamp": int(row.get("timestamp", 0)),
            "open": float(row.get("open", 0)),
            "high": float(row.get("high", 0)),
            "low": float(row.get("low", 0)),
            "close": float(row.get("close", 0)),
            "volume": float(row.get("volume", 0)),
        })
    return {
        "scenario": scenario_type.value,
        "symbol": symbol,
        "timeframe": timeframe,
        "candles": candles,
        "count": len(candles),
    }


# ── Status ───────────────────────────────────────────────────────────────────


@router.get("/simulator/status")
def simulator_status():
    eng = _get_engine()
    return {
        "active": eng.state is not None,
        "running": eng.running,
        "status": eng.state.status.value if eng.state else "IDLE",
        "session_id": eng.state.session_id if eng.state else None,
        "progress": eng._replay.progress if hasattr(eng, "_replay") else 0.0,
        "current_candle": eng._replay.index if hasattr(eng, "_replay") else 0,
        "total_candles": eng._replay.total if hasattr(eng, "_replay") else 0,
        "current_price": eng.state.current_price if eng.state else None,
        "regime": eng.state.regime.value if eng.state else "UNKNOWN",
        "trades": len(eng.state.trades) if eng.state else 0,
        "open_positions": eng.state.open_positions if eng.state else 0,
        "total_pnl": eng.state.total_pnl if eng.state else 0.0,
        "portfolio_value": eng.state.portfolio_value if eng.state else 0.0,
        "founder_mode": eng.state.config.founder_mode if eng.state else False,
        "founder_metrics": eng.state.founder_metrics if eng.state else None,
    }


# ── WebSocket ────────────────────────────────────────────────────────────────


@router.websocket("/ws/simulator")
async def ws_simulator(websocket: WebSocket):
    ws = _get_ws()
    await ws.connect(websocket, room="simulator")
    eng = _get_engine()

    try:
        while True:
            data = await websocket.receive_text()
            try:
                msg = json.loads(data)
                cmd = msg.get("command", "")
                if cmd == "get_state":
                    if eng.state:
                        await websocket.send_text(json.dumps({
                            "event": "SIMULATOR_STATE",
                            "payload": eng.state.to_dict(),
                        }))
                elif cmd == "ping":
                    await websocket.send_text(json.dumps({"event": "PONG"}))
            except json.JSONDecodeError:
                pass
    except WebSocketDisconnect:
        await ws.disconnect(websocket)
    except Exception:
        await ws.disconnect(websocket)
        raise

File: frontend/src/api/simulator.ts
import { apiFetch } from "./client";

export type SimStatus = "IDLE" | "RUNNING" | "PAUSED" | "COMPLETED" | "STOPPED";
export type SimSpeed = "1x" | "2x" | "5x" | "10x" | "100x" | "unlimited";
export type AIDecisionMode = "MANUAL" | "AI_ASSISTED" | "FULL_AI";
export type ScenarioType =
  | "FLASH_CRASH" | "BULL_RUN" | "CAPITULATION" | "RANGE"
  | "WHALE_ACCUMULATION" | "WHALE_DISTRIBUTION" | "ETF_NEWS"
  | "EXCHANGE_LISTING" | "BLACK_SWAN" | "CUSTOM";
export type MarketRegime = "BULL" | "BEAR" | "SIDEWAYS" | "VOLATILE" | "RISK_ON" | "RISK_OFF";

export interface SimulatorConfig {
  symbol: string;
  timeframe: string;
  start_date?: string;
  end_date?: string;
  initial_capital: number;
  ai_mode: AIDecisionMode;
  speed: SimSpeed;
  scenario?: ScenarioType;
  slippage_bps: number;
  fee_rate: number;
  leverage: number;
  risk_per_trade: number;
  founder_mode: boolean;
  whale_simulation: boolean;
  news_replay: boolean;
}

export interface SimulatedCandle {
  timestamp: number;
  open: number;
  high: number;
  low: number;
  close: number;
  volume: number;
}

export interface SimulatedDecision {
  id: string;
  symbol: string;
  side: string;
  timestamp: number;
  price: number;
  decision: string;
  confidence: number;
  evidence_strength: number;
  risk_score: number;
  council_report?: Record<string, unknown>;
  evidence_report?: Record<string, unknown>;
  explanation?: Record<string, unknown>;
  agent_reports: Record<string, unknown>[];
  conflicts: string[];
}

export interface SimulatedTrade {
  id: string;
  symbol: string;
  side: string;
  entry_price: number;
  entry_time: number;
  quantity: number;
  leverage: number;
  stop_loss: number;
  take_profit: number;
  trailing_stop?: number;
  status: string;
  exit_price?: number;
  exit_time?: number;
  pnl: number;
  pnl_percent: number;
  fees: number;
  slippage: number;
  close_reason?: string;
  decision_id?: string;
  elite_score?: number;
  entry_decision?: Record<string, unknown>;
  exit_decision?: Record<string, unknown>;
}

export interface TimelineEvent {
  id: string;
  timestamp: number;
  event_type: string;
  symbol: string;
  title: string;
  description: string;
  severity: string;
  data?: Record<string, unknown>;
}

export interface SimulatorState {
  session_id: string;
  status: SimStatus;
  config: SimulatorConfig;
  current_candle_index: number;
  total_candles: number;
  current_timestamp?: number;
  current_price?: number;
  elapsed_seconds: number;
  regime: MarketRegime;
  portfolio_value: number;
  cash: number;
  open_positions: number;
  total_pnl: number;
  win_count: number;
  loss_count: number;
  trades: SimulatedTrade[];
  decisions: SimulatedDecision[];
  timeline: TimelineEvent[];
  equity_curve: { timestamp: number; value: number; price: number }[];
  founder_metrics?: Record<string, number>;
}

export interface MissionReport {
  session_id: string;
  config: SimulatorConfig;
  duration_seconds: number;
  total_candles: number;
  total_decisions: number;
  total_trades: number;
  wins: number;
  losses: number;
  win_rate: number;
  total_pnl: number;
  total_fees: number;
  total_slippage: number;
  max_drawdown: number;
  max_drawdown_pct: number;
  sharpe_ratio: number;
  profit_factor: number;
  avg_win: number;
  avg_loss: number;
  largest_win: number;
  largest_loss: number;
  avg_holding_time: number;
  final_portfolio_value: number;
  return_pct: number;
  regime_changes: { timestamp: number; regime: string; title: string }[];
  elite_scores: { overall: number; confidence: number; evidence_strength: number; risk: number; execution: number; reward: number }[];
  training_score: {
    patience: number; risk: number; timing: number; entry_quality: number;
    exit_quality: number; psychology: number; discipline: number;
    missed_trades: number; mistakes: number;
  };
  trades: SimulatedTrade[];
  decisions: SimulatedDecision[];
  timeline: TimelineEvent[];
  equity_curve: { timestamp: number; value: number; price: number }[];
  mistakes: string[];
  lessons: string[];
  ai_recommendations: string[];
}

export interface SimulatorStatus {
  active: boolean;
  running: boolean;
  status: string;
  session_id?: string;
  progress: number;
  current_candle: number;
  total_candles: number;
  current_price?: number;
  regime: string;
  trades: number;
  open_positions: number;
  total_pnl: number;
  portfolio_value: number;
  founder_mode: boolean;
  founder_metrics?: Record<string, number>;
}

export interface SessionMeta {
  id: string;
  name: string;
  created_at: string;
  updated_at: string;
  config: SimulatorConfig;
  status: string;
  total_trades: number;
  total_pnl: number;
  win_rate: number;
}

export interface ScenarioInfo {
  name: string;
  description: string;
}

export function getSimulatorStatus(): Promise<SimulatorStatus> {
  return apiFetch<SimulatorStatus>("/simulator/status");
}

export function getSimulatorState(): Promise<{ status: string; state: SimulatorState | null }> {
  return apiFetch("/simulator/state");
}

export function startSimulation(config: SimulatorConfig, name = ""): Promise<{ session_id: string; state: SimulatorState }> {
  return apiFetch("/simulator/start", {
    method: "POST",
    body: JSON.stringify({ config, name }),
  });
}

export function pauseSimulation(): Promise<{ status: string }> {
  return apiFetch("/simulator/pause", { method: "POST" });
}

export function resumeSimulation(): Promise<{ status: string }> {
  return apiFetch("/simulator/resume", { method: "POST" });
}

export function stopSimulation(): Promise<{ status: string; state: SimulatorState }> {
  return apiFetch("/simulator/stop", { method: "POST" });
}

export function resetSimulation(): Promise<{ status: string }> {
  return apiFetch("/simulator/reset", { method: "POST" });
}

export function stepSimulation(): Promise<{ candle: SimulatedCandle; state: SimulatorState }> {
  return apiFetch("/simulator/step", { method: "POST" });
}

export function setSpeed(speed: SimSpeed): Promise<{ speed: string }> {
  return apiFetch(`/simulator/speed?speed=${speed}`, { method: "POST" });
}

export function seekTo(timestamp: number): Promise<{ state: SimulatorState }> {
  return apiFetch(`/simulator/seek?timestamp=${timestamp}`, { method: "POST" });
}

export function manualTrade(params: {
  side: string; entry_price: number; stop_loss: number; take_profit: number;
  quantity: number; leverage?: number; trailing_stop?: number;
}): Promise<{ trade: SimulatedTrade }> {
  const qs = Object.entries(params)
    .filter(([_, v]) => v !== undefined)
    .map(([k, v]) => `${k}=${v}`)
    .join("&");
  return apiFetch(`/simulator/trade?${qs}`, { method: "POST" });
}

export function closeTrade(tradeId: string, exitPrice?: number): Promise<{ closed: boolean }> {
  const qs = exitPrice ? `?exit_price=${exitPrice}` : "";
  return apiFetch(`/simulator/trade/${tradeId}/close${qs}`, { method: "POST" });
}

export function closeAllTrades(exitPrice?: number): Promise<{ closed_count: number }> {
  const qs = exitPrice ? `?exit_price=${exitPrice}` : "";
  return apiFetch(`/simulator/trades/close-all${qs}`, { method: "POST" });
}

export function getReport(): Promise<MissionReport> {
  return apiFetch("/simulator/report");
}

export function exportReportJson(): Promise<string> {
  return apiFetch("/simulator/report/json");
}

export function getReportPdfUrl(): string {
  return `${import.meta.env.VITE_API_URL || "http://localhost:8000"}/simulator/report/pdf`;
}

export function listSessions(): Promise<{ sessions: SessionMeta[] }> {
  return apiFetch("/simulator/sessions");
}

export function getSession(sessionId: string): Promise<SimulatorState> {
  return apiFetch(`/simulator/sessions/${sessionId}`);
}

export function deleteSession(sessionId: string): Promise<{ deleted: boolean }> {
  return apiFetch(`/simulator/sessions/${sessionId}`, { method: "DELETE" });
}

export function saveSession(sessionId: string, name = ""): Promise<{ saved: boolean }> {
  return apiFetch(`/simulator/sessions/${sessionId}/save?name=${encodeURIComponent(name)}`, { method: "POST" });
}

export function compareSessions(idA: string, idB: string): Promise<Record<string, unknown>> {
  return apiFetch(`/simulator/sessions/compare/${idA}/${idB}`);
}

export function listScenarios(): Promise<{ scenarios: Record<string, ScenarioInfo> }> {
  return apiFetch("/simulator/scenarios");
}

export function generateScenario(
  scenario_type: ScenarioType, symbol = "BTC", timeframe = "1h",
  num_candles = 200, start_price?: number
): Promise<{ scenario: string; candles: SimulatedCandle[]; count: number }> {
  const qs = `scenario_type=${scenario_type}&symbol=${symbol}&timeframe=${timeframe}&num_candles=${num_candles}${start_price ? `&start_price=${start_price}` : ""}`;
  return apiFetch(`/simulator/scenarios/generate?${qs}`, { method: "POST" });
}
